### ISSUE TITLE
Support for `@singledispatch` functions

### ISSUE THREAD

**Issue by anntzer** [sphinx-doc/sphinx#2815]
It would be nice if there was some mechanism to automagically pick up the overloads to a `@functools.singledispatch` function and list them together... e.g.

```
<Doc for the "master" function>
<links to the available overloads>
```

**Comment by tk0miya** [comment]
We have to consider how to describe (or represent) singledispatch-ed functions.

* Should we display argspec of each overload function?
* Should we display docstring of them?
* How to mention to them? link? make a list?
* Should overload functions be displayed under `automodule`?

**Comment by anntzer** [comment]
The original suggestion was to have a separate entry for each function and link them to one another, but I'm sure there are better ways to do this.

**Comment by tk0miya** [comment]
If developers uses same name for overload functions like following, it will be overrided in namespace.
```
from functools import singledispatch

@singledispatch
def func(value):
    print('default:', value)

@func.register(int)
def _(value):
    print('int:', value)

@func.register(str)
def _(value):
    print('str:', value)
```
In this case, autodoc cannot make a entry for `int` handler.

And I also worried about defining overload functions at other module:
```
from some_module import func

@func.register(object)
def _(value):
    print('object:', value)
```

In my quick investigation, it seems the overload function does not have any special attribute. So it is hard to distinguish them from normal functions.

Note: the base `singledispatch` function has special attribute: `register`, `dispatch` and `registry`. So we can pick it up easily.

**Comment by flying-sheep** [comment]
OK, so there’s 3 problems:

1. Developers often use `_` or other colliding names for registered singledispatch functions, which means they don’t stay in the namespace
2. It’s hard to detect registered singledispatch functions (but not the base function).
3. We don’t know how to represent bundles of base function + registered functions.

About 1. and 2.: We could very simply ignore the namespace here and just operate on the registry: get the set of functions in the registry that are defined in the same module as the base function and handle them together.

About 2.: In R, singledispatch is built into the language, and docs are done via manpages, so you’d see something [like this](https://stat.ethz.ch/R-manual/R-devel/library/base/html/matrix.html):

> ## Usage
> ```r
> as.matrix(x, ...)
> ## S3 method for class 'data.frame'
> as.matrix(x, rownames.force = NA, ...)
> ```
> 
> ## Arguments
> `x`
> > an R object.
> 
> `...`
> > additional arguments to be passed to or from methods.
> 
> `rownames.force`
> > logical indicating if the resulting matrix should have character (rather than NULL) rownames.

---

Combined, we could do this:

1. The user wants to document a singledispatch base function
2. We gather the set of registered functions from the same module as the base function.
3. From all signatures, we construct a “signature block”
4. We merge the registered functions’ parameter docs into the base functions’

**Comment by tk0miya** [comment]
Fixed by #7234 
Thank you for reporting.
